rules/coding-style/src/Rector/Throw_/AnnotateThrowablesRector.php

```diff
@@ -93,59 +117,77 @@ public function throwException(int $code)
         );
     }
 
-    /**
-     * @param Throw_ $node
-     */
     public function refactor(Node $node): ?Node
     {
-        if ($this->isThrowableAnnotated($node)) {
-            return null;
+        $this->throwablesToAnnotate = [];
+        if ($this->hasThrowablesToAnnotate($node)) {
+            $this->annotateThrowables($node);
+            return $node;
         }
 
-        $this->annotateThrowable($node);
-
-        return $node;
+        return null;
     }
 
-    private function isThrowableAnnotated(Throw_ $throw): bool
+    private function hasThrowablesToAnnotate(Node $node): bool
     {
-        $phpDocInfo = $this->getThrowingStmtPhpDocInfo($throw);
-        $identifiedThrownThrowables = $this->identifyThrownThrowables($throw);
+        if ($node instanceof Throw_) {
+            $this->analyzeStmtThrow($node);
+        }
 
-        foreach ($phpDocInfo->getThrowsTypes() as $throwsType) {
-            if (! $throwsType instanceof ObjectType) {
-                continue;
-            }
+        if ($node instanceof FuncCall) {
+            $this->analyzeStmtFuncCall($node);
+        }
 
-            $className = $throwsType instanceof ShortenedObjectType
-                ? $throwsType->getFullyQualifiedName()
-                : $throwsType->getClassName();
+        return count($this->throwablesToAnnotate) > 0;
+    }
 
-            if (! in_array($className, $identifiedThrownThrowables, true)) {
-                continue;
-            }
+    private function annotateThrowables(Node $node): void
+    {
+        $callee = $this->identifyCallee($node);
 
-            return true;
+        if ($callee === null) {
+            return;
         }
 
-        return false;
+        $phpDocInfo = $callee->getAttribute(AttributeKey::PHP_DOC_INFO);
+
+        foreach ($this->throwablesToAnnotate as $throwableToAnnotate) {
+            $docComment = $this->buildThrowsDocComment($throwableToAnnotate);
+            $phpDocInfo->addPhpDocTagNode($docComment);
+        }
     }
 
-    private function identifyThrownThrowables(Throw_ $throw): array
+    private function analyzeStmtThrow(Throw_ $throw): void
     {
-        if ($throw->expr instanceof New_) {
-            return [$this->getName($throw->expr->class)];
+        $foundThrownThrowables = [];
+
+        // throw new \Throwable
+        if ($throw->expr instanceof New_ && $this->getName($throw->expr->class) !== null) {
+            $foundThrownThrowables[] = $this->getName($throw->expr->class);
```

Comment:
Re-use a variable for duplicated `$this->getName($throw->expr->class)`.
It's convention to prevent multiple calls of the same code

Reply:
Done

Reply:
It is not possible to reuse the variable without writing many lines of code.

`$this->getName($throw->expr->class)` works only if `$throw->expr instanceof New_`.

The code will become unreasonably long just to reuse a variable.

If `! $throw->expr instanceof New_`. an error occurs:

> Undefined property: PhpParser\Node\Expr\MethodCall::$class

Reply:
Not sure what you mean, for me it's clear:

```php
if ($throw->expr instanceof New_) {
    $name = $this->getName($throw->expr->class);
    // ...
}
```

Reply:
I don't like, but if you think it is important, I will refactor this way:

```php
        if ($throw->expr instanceof New_) {
            $className = $this->getName($throw->expr->class);
            if (null !== $className) {
                $foundThrownThrowables[] = $this->getName($throw->expr->class);
            }
        }
```

Reply:
That's how the rest of the code base is written. Do you have idea how to make it better, explicit and robust? I'm open to learning.

<br>

Better use early return:

```php
if (null === $className) {
    continue;
}
```

Reply:
Also, I meant this way:

```diff
$className = $this->getName($throw->expr->class);
            if (null !== $className) {
-                $foundThrownThrowables[] = $this->getName($throw->expr->class);
+                $foundThrownThrowables[] = $className;
            }
```